cpp/src/aztec/honk/commitment_scheme/ipa/ipa.hpp

```diff
@@ -0,0 +1,266 @@
+        const size_t log_poly_degree = static_cast<size_t>(numeric::get_msb(poly_degree));
+        std::vector<element> L_elements(log_poly_degree);
+        std::vector<element> R_elements(log_poly_degree);
+        size_t round_size = poly_degree;
+
+        for (size_t i = 0; i < log_poly_degree; i++) {
+            round_size >>= 1;
+            // Compute inner_prod_L := < a_vec_lo, b_vec_hi > and inner_prod_R := < a_vec_hi, b_vec_lo >
+            Fr inner_prod_L = Fr::zero();
+            Fr inner_prod_R = Fr::zero();
+            for (size_t j = 0; j < round_size; j++) {
+                inner_prod_L += a_vec[j] * b_vec[round_size + j];
+                inner_prod_R += a_vec[round_size + j] * b_vec[j];
+            }
+            // L_i = < a_vec_lo, G_vec_hi > + inner_prod_L * aux_generator
+            auto pippenger_runtime_state = barretenberg::scalar_multiplication::pippenger_runtime_state(round_size);
+            element partial_L = barretenberg::scalar_multiplication::pippenger_without_endomorphism_basis_points(
+                &a_vec[0], &G_vec_local[round_size], round_size, pippenger_runtime_state);
+            partial_L += aux_generator * inner_prod_L;
+
+            // R_i = < a_vec_hi, G_vec_lo > + inner_prod_R * aux_generator
+            element partial_R = barretenberg::scalar_multiplication::pippenger_without_endomorphism_basis_points(
+                &a_vec[round_size], &G_vec_local[0], round_size, pippenger_runtime_state);
+            partial_R += aux_generator * inner_prod_R;
+
+            std::vector<element> round_commitments(2);
+            round_commitments[0] = partial_L;
+            round_commitments[1] = partial_R;
+            element::batch_normalize(&round_commitments[0], 2);
+
+            L_elements[i] = affine_element({ round_commitments[0].x, round_commitments[0].y });
+            R_elements[i] = affine_element({ round_commitments[1].x, round_commitments[1].y });
+
+            // Generate the round challenge. Todo: Use Fiat-Shamir
+            const Fr round_challenge = ipa_pub_input.round_challenges[i];
+            const Fr round_challenge_inv = round_challenge.invert();
+
+            // Update the vectors a_vec, b_vec and G_vec.
+            // a_vec_next = a_vec_lo * round_challenge + a_vec_hi * round_challenge_inv
+            // b_vec_next = b_vec_lo * round_challenge_inv + b_vec_hi * round_challenge
+            // G_vec_next = G_vec_lo * round_challenge_inv + G_vec_hi * round_challenge
+            for (size_t j = 0; j < round_size; j++) {
+                a_vec[j] *= round_challenge;
+                a_vec[j] += round_challenge_inv * a_vec[round_size + j];
+                b_vec[j] *= round_challenge_inv;
+                b_vec[j] += round_challenge * b_vec[round_size + j];
+
+                // element has (x,y,z) coord, affine_element has (x,y) coord, G_vec_local is a vector of affine_element,
+
+                auto G_lo = element(G_vec_local[j]) * round_challenge_inv;
```

Comment:
We can improve performance here by using `element::batch_mul_with_endomorphism`


This method takes a vector of input points `points` and a scalar `x` and outputs a vector containing `points[i]*x`

It's 30% faster than a basic mul operation due to performing group additions in 2D affine coordinates instead of 3D projective coordinates (affine point additions are usually more expensive than projective additions due to the need to compute a modular inverse. However we get around this by computing a single batch inverse. This only works if you are adding a lot of independent point pairs so you can amortise the cost of the single batch inversion across multiple points).


Not a priority but please add this as a TODO